zigup.zig

```diff
@@ -421,19 +425,75 @@ fn listCompilers(allocator: *Allocator) !void {
     }
 }
 
-fn printDefaultCompiler(allocator: *Allocator) !void {
+fn cleanCompilers(allocator: *Allocator) !void {
+    const install_dir_string = try getInstallDir(allocator, .{ .create = true });
+    defer allocator.free(install_dir_string);
+    // getting the current compiler
+    const default_comp_opt = try getDefaultCompiler(allocator);
+    defer if (default_comp_opt) |_| allocator.free(default_comp_opt.?);
+
+    var install_dir = std.fs.cwd().openDir(install_dir_string, .{ .iterate = true }) catch |e| switch (e) {
+        error.FileNotFound => return,
+        else => return e,
+    };
+    defer install_dir.close();
+    var it = install_dir.iterate();
+    while (try it.next()) |entry| {
+        if (entry.kind != .Directory)
+            continue;
+        if (std.mem.endsWith(u8, entry.name, ".installing"))
+            continue;
+
+        if (default_comp_opt) |default_comp| {
+            if (mem.eql(u8, default_comp, entry.name)) {
+                std.debug.warn("Not deleting the default compiler: {}\n", .{default_comp});
+                continue;
+            }
+        }
+        // if its in the compilers to keep, skip it
+
+        {
+            const look_for_keep_path = try std.fs.path.join(allocator, &[_][]const u8{ install_dir_string, entry.name, "keep" });
+            defer allocator.free(look_for_keep_path);
+            if (std.fs.openFileAbsolute(look_for_keep_path, .{ .read = false })) |keep_file| {
+                keep_file.close();
+                std.debug.warn("Not cleaning '{}' because of keep file.\n", .{entry.name});
+                continue;
+            } else |e| switch (e) {
+                error.FileNotFound => {},
+                else => return e,
+            }
+        }
+        // if we didn't continue at all then actually delete the compiler
+        const abs_path_to_delete = try std.fs.path.join(allocator, &[_][]const u8{ install_dir_string, entry.name });
```

Comment:
I think it would be better if we created this path above (before checking for keep file), then we created the `look_for_keep_path` based on this path.